File: 4/solution.cpp
class Solution {
private:
    double findNMostVal(int *arr1, int *arr2, int arr1Len, int arr2Len, int k) {
        if(arr1Len == 0) return arr2[k-1];
        else if (arr2Len == 0) return arr1[k-1];
        else if (k == 1) return min(arr1[0], arr2[0]);
        else {
            int halfIdx = k/2;
            int arr1HalfIdx = min(halfIdx, arr1Len);
            int arr2HalfIdx = min(halfIdx, arr2Len);
            if (arr1[arr1HalfIdx - 1] > arr2[arr2HalfIdx - 1]) {
                return findNMostVal(arr1, arr2+arr2HalfIdx, arr1Len, arr2Len - arr2HalfIdx, k - arr2HalfIdx);
            } else {
                return findNMostVal(arr1+arr1HalfIdx, arr2, arr1Len - arr1HalfIdx, arr2Len, k - arr1HalfIdx);
            }
        }
    }
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int *arr1 = nums1.data();
        int *arr2 = nums2.data();
        int arr1Len = nums1.size();
        int arr2Len = nums2.size();
        int totalLen = arr1Len + arr2Len;

        if (totalLen & 1) {
            // total length is odd => find the median num
            return findNMostVal(arr1, arr2, arr1Len, arr2Len, (arr1Len+arr2Len)/2 + 1);
        } else {
            // total length is even => find the avg val of two median nums
            return (findNMostVal(arr1, arr2, arr1Len, arr2Len, (arr1Len+arr2Len)/2) + findNMostVal(arr1, arr2, arr1Len, arr2Len, (arr1Len+arr2Len)/2 + 1))/2;
        }
    }
};

File: 4/solution.go
func min(a int, b int) int {
	if a > b {
		return b
	} else {
		return a
	}
}

func findKMostVal(nums1 []int, nums2 []int, m int, n int, k int) int {
	if m == 0 {
		return nums2[k-1]
	} else if n == 0 {
		return nums1[k-1]
	} else if k == 1 {
		return min(nums1[0], nums2[0])
	} else {
		halfIdx := k / 2
		mIdx := min(m, halfIdx)
		nIdx := min(n, halfIdx)

		if nums1[mIdx-1] > nums2[nIdx-1] {
			return findKMostVal(nums1, nums2[nIdx:], m, n-nIdx, k-nIdx)
		} else {
			return findKMostVal(nums1[mIdx:], nums2, m-mIdx, n, k-mIdx)
		}
	}

}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m := len(nums1)
	n := len(nums2)
	if (m+n)&1 == 1 {
		// the total length of two arr is odd
		return float64(findKMostVal(nums1, nums2, m, n, (m+n)/2+1))
	} else {
		// the total length of two arr is even
		return (float64(findKMostVal(nums1, nums2, m, n, (m+n)/2)) + float64(findKMostVal(nums1, nums2, m, n, (m+n)/2+1))) / 2
	}
}

